### PR TITLE
[Product] Scalar types in Products

### PR DESCRIPTION
**Pull request by Zales0123**
| Q               | A
| --------------- | ---
| Bug fix?        | no |
| New feature?    | no |
| BC breaks?      | yes |
| Related tickets | https://github.com/Sylius/Sylius/issues/8425 |
| License         | MIT |

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -64,6 +64,19 @@
 
 * In `PaymentInterface::setMethod` method the default value of `PaymentMethodInterface $method` parameter has been removed.
 
+### Product / ProductBundle
+
+* `ProductAttributeController` has been made final, use decoration instead of extending it.
+* `ProductVariantCombination` has been made final, use decoration instead of extending it.
+* `ProductVariantCombinationValidator
+` has been made final, use decoration instead of extending it.
```
Comment (pamil) [inline UPGRADE.md]: Something's broken here.

File: src/Sylius/Bundle/ProductBundle/Controller/ProductSlugController.php [hunk 2]
```diff
@@ -27,7 +27,7 @@ class ProductSlugController extends Controller
      *
      * @return JsonResponse
      */
-    public function generateAction(Request $request)
+    public function generateAction(Request $request): JsonResponse
```
Comment (pamil) [inline src/Sylius/Bundle/ProductBundle/Controller/ProductSlugController.php]: `Response` is enough :)

File: src/Sylius/Bundle/ProductBundle/Form/DataTransformer/ProductsToProductAssociationsTransformer.php [hunk 3]
```diff
@@ -175,7 +177,7 @@ private function setAssociatedProductsByProductCodes(ProductAssociationInterface
     /**
      * @param Collection|null $productAssociations
      */
-    private function setProductAssociations($productAssociations)
+    private function setProductAssociations(?Collection $productAssociations): void
     {
         $this->productAssociations = $productAssociations;
```
Comment (pamil) [inline src/Sylius/Bundle/ProductBundle/Form/DataTransformer/ProductsToProductAssociationsTransformer.php]: Side note: it looks weird for a transformer to have state.

File: src/Sylius/Bundle/ProductBundle/Form/Type/ProductOptionChoiceType.php [hunk 4]
```diff
@@ -52,7 +52,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
     /**
      * {@inheritdoc}
      */
-    public function configureOptions(OptionsResolver $resolver)
+    public function configureOptions(OptionsResolver $resolver): void
     {
         $resolver->setDefaults([
             'choices' => function (Options $options) {
```
Comment (pamil) [inline src/Sylius/Bundle/ProductBundle/Form/Type/ProductOptionChoiceType.php]: We can also typehint anonymous functions, `: array` here :)

File: src/Sylius/Bundle/ProductBundle/Form/Type/ProductOptionValueCollectionType.php [hunk 5]
```diff
@@ -31,14 +32,15 @@
...
-    public function buildForm(FormBuilderInterface $builder, array $options)
+    public function buildForm(FormBuilderInterface $builder, array $options): void
     {
-        if ($this->areOptionsValid($options)) {
-            throw new InvalidConfigurationException(
-                'array or (\Traversable and \ArrayAccess) of "Sylius\Component\Variation\Model\OptionInterface" must be passed to collection'
-            );
-        }
+        Assert::false(
+            $this->areOptionsValid($options),
```
Comment (pamil) [inline src/Sylius/Bundle/ProductBundle/Form/Type/ProductOptionValueCollectionType.php]: Side note: I'd move throwing the exception to that method, making it `assertOptionsAreValid` instead.

File: src/Sylius/Bundle/ProductBundle/Form/DataTransformer/ProductsToProductAssociationsTransformer.php (hunk 3 above)
Comment (Zales0123) [reply src/Sylius/Bundle/ProductBundle/Form/DataTransformer/ProductsToProductAssociationsTransformer.php]: Agreed 👍  However I would fix it in separate PR just to make this one as simple as possible.

**Comment by pamil** [comment]
Thank you very much Mateusz, astonishing work! 🐎

**Pull request merged**